nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java

```diff
@@ -0,0 +1,199 @@
+        TimeUnit queryResultTimeunit =
+            TimeUnit.valueOf(context.getProperty(INFLUX_DB_QUERY_RESULT_TIMEUNIT).evaluateAttributeExpressions(flowFile).getValue());
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        session.exportTo(flowFile, baos);
+        String query = new String(baos.toByteArray(), charset);
+
+        try {
+            long startTimeMillis = System.currentTimeMillis();
+            QueryResult result = executeQuery(context, database, query, queryResultTimeunit);
+            String json = gson.toJson(result);
+
+            if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Query result {} ", new Object[] {result});
+            }
+
+            ByteArrayInputStream bais = new ByteArrayInputStream(json.getBytes(charset));
+            session.importFrom(bais, flowFile);
+            final long endTimeMillis = System.currentTimeMillis();
+
+            if ( ! result.hasError() ) {
+                flowFile = session.putAttribute(flowFile, INFLUX_DB_EXECUTED_QUERY, String.valueOf(query));
+                session.transfer(flowFile, REL_SUCCESS);
+            } else {
+                flowFile = populateErrorAttributes(session, flowFile, query, result.getError());
+                session.transfer(flowFile, REL_FAILURE);
+            }
+
+            session.getProvenanceReporter().send(flowFile, new StringBuilder("influxdb://")
+                .append(context.getProperty(INFLUX_DB_URL).evaluateAttributeExpressions().getValue()).append("/")
+                .append(database).toString(),
+                (endTimeMillis - startTimeMillis));
+        } catch (Exception exception) {
+            flowFile = populateErrorAttributes(session, flowFile, query, exception.getMessage());
+            if ( exception.getCause() instanceof SocketTimeoutException ) {
+                getLogger().error("Failed to read from influxDB due SocketTimeoutException to {} and retrying",
+                        new Object[]{exception.getLocalizedMessage()}, exception);
+                session.transfer(flowFile, REL_RETRY);
+            } else {
+                getLogger().error("Failed to read from influxDB due to {}",
+                        new Object[]{exception.getLocalizedMessage()}, exception);
+                session.transfer(flowFile, REL_FAILURE);
+            }
+            context.yield();
+        }
+    }
+
+    protected QueryResult executeQuery(final ProcessContext context, String database, String query, TimeUnit timeunit) {
+        return getInfluxDB(context).query(new Query(query, database),timeunit);
+    }
+
+    protected FlowFile populateErrorAttributes(final ProcessSession session, FlowFile flowFile, String query,
+            String message) {
+        Map<String,String> attributes = new HashMap<>();
+        attributes.put(INFLUX_DB_ERROR_MESSAGE, String.valueOf(message));
```

Comment:
What are you using `String.valueOf` here?

Reply:
String.valueOf is a null safe method, it returns string `null` in case string is null otherwise the string, so I can safely add it to the Map.